unit_tests/recovery/main.cpp

```diff
@@ -0,0 +1,190 @@
+		<< "\t pages=" << a_database_config.max_page_count_per_volume
+		<< "\t parallel=" << a_test_config.parallel_count
+		<< "\t jobs=" << a_test_config.redo_job_count
+		<< std::endl;
+    }
+
+  cublog::redo_parallel log_redo_parallel (a_test_config.parallel_count);
+
+  ux_ut_database db_online { new ut_database (a_database_config) };
+  ux_ut_database db_recovery { new ut_database (a_database_config) };
+
+  ut_database_values_generator global_values{ a_database_config };
+  for (size_t idx = 0u; idx < a_test_config.redo_job_count; ++idx)
+    {
+      ux_ut_redo_job_impl job = db_online->generate_changes (*db_recovery, global_values);
+
+      if (job->is_volume_creation () || job->is_page_creation ())
+	{
+	  // jobs not tied to a non-null vpid, are executed in-synch
+	  db_recovery->apply_changes (std::move (job));
+	}
+      else
+	{
+	  log_redo_parallel.add (std::move (job));
+	}
+    }
+
+  log_redo_parallel.set_adding_finished ();
+  log_redo_parallel.wait_for_termination_and_stop_execution ();
+
+  db_online->require_equal (*db_recovery);
+}
+
+constexpr auto _1k = 1024u;
+constexpr auto _64k = 64 * _1k;
+constexpr auto _128k = 128 * _1k;
+
+/* small helper class to count the seconds between ctor and dtor invocations
+ */
+class measure_time
+{
+  public:
+    measure_time () = default;
+    ~measure_time ()
+    {
+      const auto end_time = std::chrono::high_resolution_clock::now ();
+      std::cout << "  duration - " << std::chrono::duration<double> (end_time - m_start_time).count () << std::endl;
+    }
+  private:
+    const std::chrono::high_resolution_clock::time_point m_start_time
+      = std::chrono::high_resolution_clock::now ();
+};
+
+/* '[ci]' tests are supposed to be executed by the Continuous Integration infrastructure
+ */
+TEST_CASE ("log recovery parallel test: some jobs, some tasks", "[ci][dbg]")
+{
+  srand (time (nullptr));
+  initialize_thread_infrastructure ();
+
+  std::array<size_t, 1> volume_count_per_database_arr { 10u };
+  std::array<size_t, 1> page_count_per_volume_arr { _1k };
+  std::array<size_t, 2> job_count_arr { 0u, _64k };
+  std::array<size_t, 2> parallel_count_arr { 1u, std::thread::hardware_concurrency ()};
+  for (const size_t volume_count_per_database : volume_count_per_database_arr)
+    for (const size_t page_count_per_volume : page_count_per_volume_arr)
+      for (const size_t job_count : job_count_arr)
+	for (const size_t parallel_count : parallel_count_arr)
+	  {
+	    const log_recovery_test_config test_config =
+	    {
+	      parallel_count, // std::thread::hardware_concurrency (), // parallel_count
+	      job_count, // redo_job_count
+	      false, // verbose
+	    };
+
+	    const ut_database_config database_config =
+	    {
+	      volume_count_per_database, // max_volume_count_per_database
+	      page_count_per_volume, // max_page_count_per_volume
+	      0., // max_duration_in_millis
+	    };
+
+	    execute_test (test_config, database_config);
+	  }
```

Comment:
Add the scopes for each for loop, similar to the long test case.

Reply:
Done